### ISSUE TITLE
[Feature Request]: Introduce platform parameter dashboard (GSoC'25 4.2)

### ISSUE THREAD

**Issue by BenHenning** [oppia/oppia-android#5345]
### Is your feature request related to a problem? Please describe.

Feature flags are a special type of configurable [platform parameter](https://github.com/oppia/oppia-android/wiki/Platform-Parameters-&-Feature-Flags#introduction) which allows the team to stage features behind remotely configurable flags until they're ready to be launched. This allows features to be developed across multiple releases without users seeing part of the feature (or app stability issues when the feature is enabled), ensuring the team releases high-quality features and doesn't hurt the overall quality and performance of the app. Broadly, platform parameters allow the team to configure the overall app (which can be useful both for feature flags, as described above, and safety 'knobs' such as controlling rate limits to remote APIs to help reduce the chance of server outages).

This project entails introducing a developer-only UI (as part of the developer options section of the app) which displays all platform parameters and feature flags in the app, their current enabled/disabled status (for feature flags) or values (for platform parameters), and their sync status (i.e. whether they're being synced from the server or using a local developer default). It also allows an explicit manual override to force the feature on or off, or to override the platform parameter's value.

Note that implementations should be based on the new version of platform parameters being introduced in https://github.com/oppia/oppia-android/pull/5725.

#### Suggested Milestones:
- **Milestone 1**:
  - Display a list of platform parameters and feature flags from a Developer Options menu, along with their current values and sync statuses.
  - Set up initial tests to demonstrate that the UI displays correctly.
  - The new UI correctly shows all platform parameters and feature flags, and their correct values and sync statuses.
  - Support for overriding feature flag values (requires manual restart).
- **Milestone 2**:
  - Support for overwriting and resetting both feature flags and platform parameters back to their default values, including force-restarting the app upon navigating away from the menu (so that the changes can take effect).
  - Updated UI tests for the new functionality.
  - The new screen fully supports overriding various platform parameters and feature flag values.
  - Pending overrides should 'survive' device rotations (that is, rotating the device should not cause any temporary state in the UI to be lost).
  - Support for force downloading flags from Oppia web (i.e. by calling PlatformParameterController.downloadRemoteParameters() and asking for an app restart). 
  - The indicator for downloading state should 'survive' device rotations.
  - Redesign the dashboard in accordance with the latest mockup layout - [Figma Link](https://www.figma.com/proto/Z09eMDliYwgjHCOIvQrBDI/Oppia-Platform-Parameters---Feature-Flags---Developer-Options?node-id=653-7203&t=WjxWDKnv3P7TCmab-0&scaling=min-zoom&content-scaling=fixed&page-id=653%3A1993&starting-point-node-id=653%3A5660).

- **PM Demo Changes**
  - Input Validation: Edit text in the platform parameter dashboard should allow at least 6 characters/digits.
  - Invalid Input Handling: On losing focus with invalid input, the field should revert to the last valid value.
  - Unsaved Changes Indicator: Unsaved manual edits should be highlighted with a blue indicator.
  - Save Button Behavior: A Save button at the top-right should toggle enabled/disabled based on pending changes.
  - Back Navigation Flow: On back navigation:
    - If there are pending changes, show a Save/Discard dialog.
      - Save → saves changes and prompts restart.
      - Discard → reverts overrides and returns to developer options.
      - Clicking outside → keeps the user on the dashboard.

<details>
<summary>Technical hints / guidance</summary>

- General changes explanation:
  - Much of the new UI is fairly straightforward to implement, and both the sync status and values are well represented in protos.
  - The main complexity comes in providing clean pathways for overriding values, and retaining these values across app restarts.
- UI changes:
  - New developer options view model & listener for representing the platform parameter/feature flag dashboard.
  - New modal fragment/presenter and listener for supporting back navigation and confirming an app restart after making an override change.
  - New fragment, presenter, view model, listeners, and RecyclerView view models + views for the actual platform parameter and feature flag list screens themselves, plus changes for override functionality.
- Domain changes:
  - Introduce a new implementation of PlatformParameterController specific to developer options (similar to how HintHandlerDebugImpl exists and is used). This controller:
    - Delegates functionality to the production implementation of the controller, but overrides as necessary.
    - Exposes the necessary functionality for displaying platform parameters and feature flags.
    - Supports local override for syncing by:
      - Introducing a new persistent cache store for the LocalOverridePlatformParameterDatabase mentioned below).
      - Introducing a new method for computing the list of ephemeral platform parameters by using both the remote sync and local override, with a resolution order of: pick local override if present, pick remote if present, fallback to default otherwise. This should build on the existing functionality provided by the production implementation.
    - Introduce an update function for locally overriding platform parameters that changes the new database (but should not affect current in-process state set in PlatformParameterProcessState).
    - Introduce support for resetting a flag/parameter (which should just remove its entry from the override database).
- Model changes:
  - A new LocalOverridePlatformParameterDatabase proto (essentially mirroring the existing RemotePlatformParameterAndFeatureFlagDatabase proto) for storing overridden values. This is kept separate from the other database so that a flag/parameter 'reset' can restore back to the original sync state.
  - Add 'LOCAL_OVERRIDE' to sync status. This should be used in the override database by default so that the UI is able to properly display the badge for a locally overridden value.
- Testing hints:
  - Refer to [this example](https://github.com/oppia/oppia-android/blob/08e00847d955ec1f38fe8000832a2cff1bbff62f/app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt#L1043) to simulate an app restart in tests. Testing should ensure that overridden values are retained across restarts.
</details>

### Additional context

Note that this project includes or relates to work from following issues: #4302 and #4303.

**Mockups:**

| Feature | Mockup |
|---------|--------|
| Platform Parameter Dashboard & Force Downloading Server Parameters | [Figma Link](https://www.figma.com/proto/Z09eMDliYwgjHCOIvQrBDI/Oppia-Platform-Parameters---Feature-Flags---Developer-Options?node-id=653-7203&t=WjxWDKnv3P7TCmab-0&scaling=min-zoom&content-scaling=fixed&page-id=653%3A1993&starting-point-node-id=653%3A5660)|

 <details><summary>Outdated Mockup</summary><br> 

- Platform Parameter Dashboard | [Mockup](https://www.figma.com/proto/Z09eMDliYwgjHCOIvQrBDI/Oppia-Platform-Parameters-%2F-Feature-Flags---Developer-Options?node-id=51-618&t=Nj1FSaojE8Roe9Dk-0&scaling=min-zoom&content-scaling=fixed&page-id=51%3A334&starting-point-node-id=51%3A618)
- Force Downloading Server Flags| [Mockup](https://www.figma.com/proto/Z09eMDliYwgjHCOIvQrBDI/Oppia-Platform-Parameters---Feature-Flags---Developer-Options?node-id=311-441&t=cw3Kuy9Xlab0eFVO-1&scaling=min-zoom&content-scaling=fixed&page-id=51%3A334&starting-point-node-id=51%3A618)  
    *(Note: The design isn't finalized yet and may evolve, but this serves as a reference for understanding the placement and expected behavior of the Force Download option.)*

</details> 

### Noted Gaps:
Note: There are a few gaps in the project, and the scope has been adjusted accordingly, with issues filed as listed below.
- The actual working of the force download functionality—which is currently limited to the UI due to the unavailability of an API to handle it. [Issue #5950]
- The inability to display the last synced timestamps on server sync. [Issue #5951]

### Additional Resources:
- Meeting Minutes : [Doc Link](https://docs.google.com/document/d/1HscU1eKhoDVFfbJeXbTy_g8_T2qCC9wZYNC9ghNm1Ac/edit?tab=t.0#heading=h.n4eokff22o9r)
- GSoC Weekly Journal : [Journal Link](https://medium.com/@theayushyadav11)

### Milestone Table

# ✅ Milestone 1

### ✅ PR 1.1 - [[5845](https://github.com/oppia/oppia-android/pull/5845)]  
**Implementing Debug Platform Parameter Controller**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 02-06-2025    | 05-06-2025     | 08-06-2025 |

| Status | Tasks                                                                                     | Due Date     |
|--------|-------------------------------------------------------------------------------------------|--------------|
| ✅     | Introduce proto definitions for `EphemeralFeatureFlag` and `EphemeralPlatformParameter`.   | 02-06-2025   |
| ✅     | Implement `PlatformParameterDebugModule` to provide the debug controller.                  | 03-06-2025   |
| ✅     | Implement `PlatformParameterControllerDebugImpl` with load & download functions.           | 04-06-2025   |
| ✅     | Write controller logic to resolve and provide the Ephemeral Feature Flags.                 | 04-06-2025   |
| ✅     | Write controller logic to resolve and provide the Ephemeral Platform Parameters.           | 04-06-2025   |
| ✅     | Write tests for the Controller.                                                             | 05-06-2025   |

<h1></h1>

### ✅ PR 1.2 - [[5861](https://github.com/oppia/oppia-android/pull/5861)]  
**Introduce Feature Flag Dashboard**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 06-06-2025    | 13-06-2025     | 17-06-2025 |

| Status | Tasks                                                                                     | Due Date     |
|--------|-------------------------------------------------------------------------------------------|--------------|
| ✅     | Add menu item and listener for Feature Flag screen.                                       | 06-06-2025   |
| ✅     | Create `FeatureFlagActivity`, presenter and layout.                                       | 08-06-2025   |
| ✅     | Create `FeatureFlagFragment`, presenter, ViewModel and layout.                            | 11-06-2025   |
| ✅     | Wire controller with ViewModel and UI for dataflow.                                       | 12-06-2025   |
| ✅     | Handle configuration change persistence.                                                  | 12-06-2025   |
| ✅     | Write tests for feature flags.                                                            | 13-06-2025   |

<h1></h1>

### ✅ PR 1.3 - [[5867](https://github.com/oppia/oppia-android/pull/5867)]  
**Introduce Platform Parameter Dashboard**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 14-06-2025    | 23-06-2025     | 27-06-2025 |

| Status | Tasks                                                                                     | Due Date     |
|--------|-------------------------------------------------------------------------------------------|--------------|
| ✅     | Add menu item and listener for Platform Parameter screen.                                 | 14-06-2025   |
| ✅     | Create `PlatformParameterActivity`, presenter and layout.                                 | 16-06-2025   |
| ✅     | Create `PlatformParameterFragment`, presenter, ViewModel and layout.                      | 20-06-2025   |
| ✅     | Wire controller with ViewModel and UI for dataflow.                                       | 22-06-2025   |
| ✅     | Handle configuration change persistence.                                                  | 22-06-2025   |
| ✅     | Write UI and controller tests for Platform Parameters.                                    | 23-06-2025   |

<h1></h1>

### ✅ PR 1.4 - [[5871](https://github.com/oppia/oppia-android/pull/5871)]  
**Implement Feature Flag overrides & loading override behavior**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 25-06-2025    | 02-07-2025     | 04-07-2025 |

| Status | Tasks                                                                                     | Due Date     |
|--------|-------------------------------------------------------------------------------------------|--------------|
| ✅       | Create `LocalOverridePlatformParameterDatabase` proto & update `SyncStatus` enum.         | 24-06-2025   |
| ✅       | Implement controller to resolve flags with overridden + remote + default.                 | 25-06-2025   |
| ✅       | Implement `updateOverriddenFeatureFlag()` in the controller.                              | 27-06-2025   |
| ✅       | Update UI in `FeatureFlagFragmentPresenter` and `FeatureFlagViewModel`.                   | 30-06-2025   |
| ✅      | Implement loading logic using overridden values.                                           | 01-07-2025   |
| ✅       | Write comprehensive UI and controller tests.                                               | 02-07-2025   |

<h1></h1>

# ✅ Milestone 2

### ✅ PR 2.1 - [[5892](https://github.com/oppia/oppia-android/pull/5892)]  
**Implement Platform Parameter overrides & loading behavior**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 26-07-2025    | 02-08-2025     | 06-08-2025 |

| Status | Tasks                                                                                     | Due Date     |
|--------|-------------------------------------------------------------------------------------------|--------------|
| ✅      | Implement controller to resolve parameters with overridden + remote + default.            | 26-07-2025   |
| ✅      | Implement `updateOverriddenPlatformParameter()` in the controller.                        | 28-07-2025   |
| ✅      | Update UI in `PlatformParameterFragmentPresenter` and `PlatformParameterViewModel`.       | 30-07-2025   |
| ✅      | Implement loading logic using overridden values.                                           | 01-08-2025   |
| ✅      | Write comprehensive UI and controller tests.                                               | 02-08-2025   |

<h1></h1>

### ✅ PR 2.2 - [[5894](https://github.com/oppia/oppia-android/pull/5894)]  
**Implement reset functionality for overrides**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 06-08-2025    | 09-08-2025     | 18-08-2025 |

| Status | Tasks                                                                                     | Due Date     |
|--------|-------------------------------------------------------------------------------------------|--------------|
| ✅     | Implement `resetPlatformParameterOverride()` & `resetFeatureFlagOverride()`.              | 06-08-2025   |
| ✅     | Update UI for reset in both feature flag & platform parameter fragments.                   | 08-08-2025   |
| ✅     | Implement the new UI design changes.                  | 13-08-2025   |
| ✅     | Write tests to verify reset functionality.                                                | 14-08-2025   |

<h1></h1>

### ✅ PR 2.3 - [[5929](https://github.com/oppia/oppia-android/pull/5929)]  
**Add restart dialogs & logic for overrides/reset**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 13-08-2025    | 17-08-2025     | 21-08-2025 |

| Status | Tasks                                                                                         | Due Date     |
|--------|-----------------------------------------------------------------------------------------------|--------------|
| ✅     | Implement `isAppRestartRequiredForFeatureFlag()` & `isAppRestartRequiredForPlatformParameter()` | 14-08-2025   |
| ✅     | Create `ConfirmRestartDialogFragment`, presenter and layout.                                  | 15-08-2025   |
| ✅     | Implement forced app restart logic.                                                           | 16-08-2025   |
| ✅    | Write tests for restart functionality.                                                        | 17-08-2025   |

<h1></h1>

### ✅ PR 2.4 - [[5949](https://github.com/oppia/oppia-android/pull/5949/)]  
**Introduce Force Remote Download Support**

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 18-08-2025    | 24-08-2025     | 27-08-2025 |

| Status | Tasks                                                                                     | Due Date     |
|--------|-------------------------------------------------------------------------------------------|--------------|
| ✅     | Introduce controller `ForceDownloadParameterController`.                                  | 18-08-2025   |
| ✅     | Add menu item for force download and its listener.                                        | 19-08-2025   |
| ✅     | Create `ForceDownloadDialogFragment`, presenter and layout.                               | 20-08-2025   |
| ✅     | Handle logic for force downloading, config change handling.                               | 22-08-2025   |
| ✅     | Implement cancel, restart, and controller logic.                                         | 23-08-2025   |
| ✅     | Write UI and controller tests for force downloading.                                      | 24-08-2025   |

<h1></h1>

### ✅ Evaluation Fix PR 1 - [[5953](https://github.com/oppia/oppia-android/pull/5953/)]  
**Pull request to resolve evaluation-blocking issues**

| Creation Date | 
|----------------|
| 06-09-2025     | 

| Status | Tasks                                                                                     |  
|--------|-------------------------------------------------------------------------------------------|
| ✅     | Fix restarting of app on orientation changes                                 | 

### ✅ Evaluation Fix PR  2- [[5958](https://github.com/oppia/oppia-android/pull/5958/)]  
**Pull request to resolve evaluation-blocking issues**

| Creation Date | 
|----------------|
| 09-09-2025     | 

| Status | Tasks                                                                                     |  
|--------|-------------------------------------------------------------------------------------------|
| ✅     | Fix restarting of download on orientation changes                                 |

**Comment by BenHenning** [comment]
NB: This issue is locked to ensure that it only contains updates for requirements. If you have questions regarding this project, please use its corresponding discussion category: https://github.com/oppia/oppia-android/discussions/categories/gsoc-q-a-4-android-projects.

**Comment by BenHenning** [comment]
NB: The main issue comment has been updated to include the technical details directly rather than a link to a Gist (so that we can easily change this if needed). None of the technical requirements have actually changed.

**Comment by BenHenning** [comment]
Please note that this project is no longer being offered as a Google Summer of Code 2024 project for Oppia Android since we can't be sure that we'll have a mentor for it.

Existing proposals that have been submitted for feedback will still be reviewed, and the corresponding authors will be met with to discuss options.

**Comment by theayushyadav11** [comment]
# Milestone 1  
### ⏳ PR 1.1 - [#]()
Introduce the model protos required for the project (LocalOverridePlatformParameterDatabase).

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 02-06-2025    | 03-06-2025     | 04-06-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Create LocalOverridePlatformParameterDatabase proto definition.  | 02-06-2025   |
| ⏳| Add necessary message structures for platform parameters and feature flags.  | 03-06-2025   |
| ⏳| Test proto compilation and integration.  | 04-06-2025   |

### ⏳ PR 1.2 - [#]()
Introduce the PlatformParameterControllerDebugImpl and module to inject it.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 05-06-2025    | 06-06-2025     | 12-06-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Implement PlatformParameterControllerDebugImpl class.  | 06-06-2025   |
| ⏳| Create dependency injection module for debug controller.  | 08-06-2025   |
| ⏳| Integrate controller with existing platform parameter system.  | 10-06-2025   |
| ⏳| Add unit tests for controller implementation.  | 12-06-2025   |

### ⏳ PR 1.3 - [#]()
Introduce menu item, listener, controller, dialog fragment for Force Downloading.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 09-06-2025    | 10-06-2025     | 14-06-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Add menu item for force download functionality and its listener.  | 10-06-2025   |
| ⏳| Implement `ForceDownloadParameterController`.  | 11-06-2025   |
| ⏳| Create dialog fragment for force download confirmation.  | 12-06-2025   |
| ⏳| Add download state indicator with rotation survival.  | 13-06-2025   |
| ⏳| Test force download functionality.  | 14-06-2025   |

### ⏳ PR 1.4 - [#]()
Implement Feature Flag Screen functionality.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 11-06-2025    | 12-06-2025     | 21-06-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Add menu item and listener for Feature Flag screen.  | 12-06-2025   |
| ⏳| Create `FeatureFlagActivity` its presenter and its layout.  | 15-06-2025   |
| ⏳| Create `FeatureFlagFragment` its presenter, ViewModel and its layout.  | 16-06-2025   |
| ⏳| Integrate Feature Flag screen with controller.  | 17-06-2025   |
| ⏳| Write comprehensive tests for Feature Flags functionality.  | 19-06-2025   |
| ⏳| Handle configuration changes and state persistence.  | 21-06-2025   |

### ⏳ PR 1.5 - [#]()
Implement Platform Parameter Screen functionality.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 20-06-2025    | 21-06-2025     | 28-06-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Add menu item and listener for Platform Parameter screen.  | 21-06-2025   |
| ⏳|  Create `PlatformParameterActivity` its presenter and its layout.  | 23-06-2025   |
| ⏳| Create `PlatformParameterFragment` its presenter, ViewModel and its layout. | 24-06-2025   |
| ⏳| Integrate Platform Parameter screen with controller.  | 25-06-2025   |
| ⏳| Write comprehensive tests for Platform Parameters functionality.  | 27-06-2025   |
| ⏳| Handle configuration changes and state persistence.  | 28-06-2025   |

### ⏳ PR 1.6 - [#]()
Final milestone 1 tasks and PM feedback implementation.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 29-06-2025    | 01-07-2025     | 03-07-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Implement PM suggestions and feedback.  | 01-07-2025   |
| ⏳| Final integration testing and bug fixes.  | 02-07-2025   |
| ⏳| Prepare demo with flag list, values, and force downloading.  | 03-07-2025   |

# Milestone 2  
### ⏳ PR 2.1 - [#]()
Implement Feature Flag overrides functionality.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 26-07-2025    | 26-07-2025     | 12-08-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Make proto changes for feature flag overrides.  | 26-07-2025   |
| ⏳| Implement controller changes for feature flag overrides.  | 30-07-2025   |
| ⏳| Create UI and domain flow for feature flag overrides.  | 04-08-2025   |
| ⏳| Implement configuration change survival for overrides.  | 06-08-2025   |
| ⏳| Write comprehensive tests (unit, integration, edge cases).  | 10-08-2025   |
| ⏳| Code review and bug fixes.  | 12-08-2025   |

### ⏳ PR 2.2 - [#]()
Implement Platform Parameter overrides functionality.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 08-08-2025    | 09-08-2025     | 22-08-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Make proto changes for platform parameter overrides.  | 09-08-2025   |
| ⏳| Implement controller changes for platform parameter overrides.  | 12-08-2025   |
| ⏳| Implement configuration change survival for parameters.  | 14-08-2025   |
| ⏳| Create UI and domain flow for platform parameter overrides.  | 17-08-2025   |
| ⏳| Write comprehensive tests (unit, integration, edge cases).  | 20-08-2025   |
| ⏳| Code review and integration testing.  | 22-08-2025   |

### ⏳ PR 2.3 - [#]()
Implement Reset functionality for both Feature Flags and Platform Parameters.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 21-08-2025    | 22-08-2025     | 24-08-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Implement reset logic for feature flags to original values.  | 22-08-2025   |
| ⏳| Implement reset logic for platform parameters to defaults.  | 23-08-2025   |
| ⏳| Add persistence handling for reset functionality.  | 24-08-2025   |

### ⏳ PR 2.4 - [#]()
Add "Restart" dialogs for overrides and reset implementation.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 23-08-2025    | 24-08-2025     | 25-08-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Create restart dialog `ConfirmRestartDialogFragment`, its presenter and layout. | 24-08-2025   |
| ⏳| Implement forced app restart logic for applying changes.  | 25-08-2025   |
| ⏳| Test restart functionality with overrides and resets.  | 25-08-2025   |

### ⏳ PR 2.5 - [#]()
Final milestone 2 tasks and PM feedback implementation.

| Starting Date | Creation Date | Merge Date |
|---------------|----------------|-------------|
| 25-08-2025    | 26-08-2025     | 28-08-2025 |

|      | Tasks                                                       | Due Date     |
|----|-------------------------------------------------------------|--------------|
| ⏳| Implement PM suggestions and comments.  | 26-08-2025   |
| ⏳| Final integration testing and bug fixes.  | 27-08-2025   |
| ⏳| Prepare demo showing override and reset functionality.  | 28-08-2025   |

**Comment by Nik-09** [comment]
Hi @theayushyadav11, could you please add the milestone table directly to the PR description? It’ll help ensure it doesn’t get lost in the thread as the conversation grows. Thanks!

**Comment by Rd4dev** [comment]
@theayushyadav11, could you remove the table for now and add it back once it’s fully completed and approved? Since it’s still a work in progress and several comments remain unaddressed, it would be helpful if you could provide an ETA for completion. Please make sure to address all feedback and assign it for review once done, so we can finalize and include the accurate version.

@Nik-09, also requesting this since the milestone table still needs some refinement — but happy to hear your thoughts on how you'd prefer to proceed here. [Ref: [Milestone Table Doc](https://docs.google.com/document/d/1tgozURrIvNS5xtPocMTq0x59gadh_Di9TKuKD_DZz-E/edit?usp=sharing)]

**Comment by Rd4dev** [comment]
## Guide to PR Chaining

You should have write permission to the repository to proceed with PR chaining — which you likely should have been already granted by the leads, so no action is needed there.

### Setting Up `upstream` Remote

Make sure your `upstream` is set to the **oppia/oppia-android** repository.
To verify:

```bash
$ git remote -v
```

It should list something like:

```
origin    https://github.com/username/oppia-android.git (fetch)
origin    https://github.com/username/oppia-android.git (push)
upstream  https://github.com/oppia/oppia-android.git (fetch)
upstream  https://github.com/oppia/oppia-android.git (push)
```

If `upstream` isn’t set, run:

```bash
$ git remote add upstream https://github.com/oppia/oppia-android.git
```

<h1></h1>

### Pushing Your Local Branch

Once you've committed your changes on your local branch, do the initial upstream push:

```bash
$ git push --set-upstream upstream your-branch-name
```

This is a **one-time setup** for the branch to let Git know to track it against the `upstream` remote. After this, you can simply run `git push` as usual.

> This is required because **PR chaining only works when branches are pushed to `upstream`**, not to your fork.

<h1></h1>

### 🔗 Creating Linked PRs

When creating a PR, set the **base branch** as the branch you based off of. This ensures the PR includes only the relevant changes for that feature.

>Additional Context:
If the **first PR** is branched directly from `develop` and doesn’t depend on any other PRs, then `develop` should be the base branch. But if your first PR is **based on another feature branch**, then that feature branch should be used as the base.
Ideally, the starting point for the project (i.e., the very first independent PR) should branch off from `develop`, and then each subsequent PR can chain off the one before it.

![Image](https://github.com/user-attachments/assets/1464bf31-3ebc-4f64-936d-8a0349c781fb)

<br> 
Eventually, the PR base will automatically change to `develop` once the preceding PR is merged.
Also, always keep your feature branches updated especially after the chained PR is merged into develop.
<br/> <br/>

>For example: 
These PRs represent how multiple PRs are linked in sequence — #4253, #4257, #4258, #4262.
![Image](https://github.com/user-attachments/assets/b0391543-b057-4e54-949d-596c6810650c)
![Image](https://github.com/user-attachments/assets/35283a61-894b-4e57-a6fd-60d37a4eae29)
![Image](https://github.com/user-attachments/assets/d699c185-f217-4067-bf8e-5be2a19dd4a3)
![Image](https://github.com/user-attachments/assets/f6557649-186e-497d-9399-7bec53976ce6)

this structure will be followed for medium/large tasks moving forward.
> For Unlinked PRs -- you can / should push using your fork (not upstream)

<h1></h1>


### 🚫 Caution

If you’re pushing directly to the main `oppia-android` repo (i.e., `upstream`), **never commit binary files or as excessively large PRs directly**.

**Comment by BenHenning** [comment]
This feature has been completed and launched.
